compiler/rustc_parse/src/parser/item.rs

```diff
@@ -81,6 +81,11 @@ impl<'a> Parser<'a> {
                         span,
                         "consider using `const` or `static` instead of `let` for global variables",
                     );
+                } else if self.parse_const_block(span, false).is_ok() {
```

Comment:
Using `is_ok` here is kinda hazardous since you need to cancel the diagnostic on the bad path.

Reply:
that is... definitely the part I was least certain about, yes. is there, like, a sample of Rust that will now elicit some disastrously bad diagnostic as a result of doing this, or...?

Reply:
No, if you drop a `Diagnostic` without emitting it the compiler will ICE.

Presumably you will have some code that looks like:

```rust
match self.parse_const_block(span, false) {
  Err(e) => { e.cancel(); }
  Ok(_) => { do the recovery code here }
}
```

Which at that point we probably should do snapshotting too, so try this:

```rust
let snapshot = self.create_snapshot_for_diagnostic();
match self.parse_const_block(span, false) {
  Err(e) => { 
       e.cancel();
       self.recover_from_snapshot(snapshot);
   }
  Ok(_) => { do the recovery code here }
}
```

Reply:
> No, if you drop a Diagnostic without emitting it the compiler will ICE.

...oh lmao, drop bombs away.

Reply:
Example of the drop bomb in question -- this ICEs after your PR:

```
const { a: () = 1; };
```